test/iaas.GcpClient.spec.js

```diff
@@ -0,0 +1,538 @@
+      });
+      it('should throw an error if credentials object provided in config is missing any params', function () {
+        var config = {
+          name: 'gcp',
+          projectId: 'my-project-id',
+          credentials: {
+            'type': 'service_account',
+            'project_id': 'my-gcp-dev',
+            'private_key_id': 'kkkkkkkk',
+            'private_key': '-----BEGIN PRIVATE KEY-----\nlllllllllllllllllllllll\n-----END PRIVATE KEY-----\n',
+            'client_email': 'my-test-user@my-gcp-dev.iam.myaccounts.com',
+            'client_id': '000000000000000',
+            'auth_uri': 'https://myaccounts.com/oauth2/auth',
+            'token_uri': 'https://myaccounts.com/oauth2/token',
+            'auth_provider_x509_cert_url': 'https://myaccounts.com/oauth2/v1/certs',
+            // missing client_x509_cert_url param
+          }
+        };
+        expect(() => GcpClient.validateParams(config)).to.throw();
+      });
+      it('should return true if projectId and credentials are provided in config', function () {
+        expect(GcpClient.validateParams(settings)).to.equal(true);
+      });
+    });
+
+    describe('#GcpStorage', function () {
+      it('should form an object with credentials and projectId with storage base url', function () {
+        const responseGcpStorageObject = GcpClient.createStorageClient(settings);
+        const baseUrl = 'https://www.googleapis.com/storage/v1';
+
+        expect(responseGcpStorageObject.baseUrl).to.equal(baseUrl);
+        expect(responseGcpStorageObject.projectId).to.equal(settings.projectId);
+        expect(responseGcpStorageObject.authClient.config.projectId).to.equal(settings.projectId);
+        expect(responseGcpStorageObject.authClient.config.credentials).to.eql(settings.credentials);
+      });
+    });
+
+    describe('#GcpCompute', function () {
+      it('should form an object with credentials and projectId with compute base url', function () {
+        const responseGcpComputeObject = GcpClient.createComputeClient(settings);
+        const baseUrl = 'https://www.googleapis.com/compute/v1';
+
+        expect(responseGcpComputeObject.baseUrl).to.equal(baseUrl);
+        expect(responseGcpComputeObject.projectId).to.equal(settings.projectId);
+        expect(responseGcpComputeObject.authClient.config.credentials).to.eql(settings.credentials);
+      });
+    });
+
+    describe('#BlobOperations', function () {
+      before(function () {
+        sandbox.stub(GcpStorage.prototype, 'bucket').withArgs(settings.container).returns(bucketStub);
+        sandbox.stub(utils, 'streamToPromise', utilsStreamToPromiseStub);
+      });
+      after(function () {
+        sandbox.restore();
+      });
+
+      it('container properties should be retrived successfully', function () {
+        return client.getContainer()
+          .then(result => {
+            expect(result[1].name).to.equal(settings.container);
```

Comment:
should the getContainer method return an array or a single object? Since you are fetching a particular container wouldnt it be better that the method returns `result[0]` so that upstream always knows its container details? Also check from other IAAS clients if this method returns an object or an array. So keep it in line.

Also you are referring as `result[1]` & `result[0]` is put as empty object. Why is this?

Reply:
getContainer call from GCP returns Bucket object in `result[0]` and metadata in `result[1]` as written in [Bucket#get](https://cloud.google.com/nodejs/docs/reference/storage/1.6.x/Bucket#get). So, instead of returning the array of result, we can return the metadata in `result[1]`

Reply:
Even AzureClient returns metadata. So, keeping the same behaviour for GCPClient as well.